File: code/packages/haskell/wave/src/Wave.hs
-- | A validated simple-harmonic wave model.
module Wave
    ( Wave
    , newWave
    , newWaveWithPhase
    , amplitude
    , frequency
    , phase
    , period
    , angularFrequency
    , evaluate
    ) where

import qualified Trig
import Data.Bits (shiftL)

-- | A sinusoidal wave whose amplitude and frequency have been validated.
data Wave = Wave
    { amplitude :: Double
    , frequency :: Double
    , phase :: Double
    } deriving (Eq, Show)

-- | Construct a wave with zero phase.
newWave :: Double -> Double -> Either String Wave
newWave waveAmplitude waveFrequency =
    newWaveWithPhase waveAmplitude waveFrequency 0.0

-- | Construct a wave with an explicit phase in radians.
newWaveWithPhase :: Double -> Double -> Double -> Either String Wave
newWaveWithPhase waveAmplitude waveFrequency wavePhase
    | not (isFinite waveAmplitude && isFinite waveFrequency && isFinite wavePhase) =
        Left "wave parameters must be finite"
    | waveAmplitude < 0.0 = Left "amplitude must be non-negative"
    | waveFrequency <= 0.0 = Left "frequency must be positive"
    | waveFrequency > maxFinite / Trig.twoPi = Left "angular frequency must be finite"
    | otherwise = Right Wave
        { amplitude = waveAmplitude
        , frequency = waveFrequency
        , phase = wavePhase
        }

-- | Duration of one complete cycle in seconds.
period :: Wave -> Double
period wave = 1.0 / frequency wave

-- | Angular frequency in radians per second.
angularFrequency :: Wave -> Double
angularFrequency wave = 2.0 * Trig.piValue * frequency wave

-- | Evaluate the wave at time @t@ in seconds.
evaluate :: Wave -> Double -> Double
evaluate wave time
    | not (isFinite time) = error "time must be finite"
    | amplitude wave == 0.0 = 0.0
    | otherwise = amplitude wave * clampedUnit
  where
    reducedTime = signedRemainder time (period wave)
    reducedPhase = signedRemainder (phase wave) Trig.twoPi
    angle = Trig.twoPi * (frequency wave * reducedTime) + reducedPhase
    clampedUnit = max (-1.0) (min 1.0 (Trig.sin angle))

maxFinite :: Double
maxFinite = 1.7976931348623157e308

isFinite :: Double -> Bool
isFinite value = not (isNaN value || isInfinite value)

-- | Base-only signed remainder that never forms the potentially overflowing
-- quotient @value / modulus@. Both inputs are represented exactly as an
-- Integer mantissa times a power of two, aligned, and reduced with 'rem'.
signedRemainder :: Double -> Double -> Double
signedRemainder value modulus
    | isInfinite modulus = value
    | value == 0.0 = value
    | otherwise =
        encodeFloat (scaledValue `rem` scaledModulus) commonExponent
  where
    (valueMantissa, valueExponent) = decodeFloat value
    (modulusMantissa, modulusExponent) = decodeFloat modulus
    commonExponent = min valueExponent modulusExponent
    scaledValue = valueMantissa `shiftL` (valueExponent - commonExponent)
    scaledModulus = modulusMantissa `shiftL` (modulusExponent - commonExponent)

File: code/packages/dart/wave/lib/src/wave.dart
import 'package:coding_adventures_trig/trig.dart' as trig;

/// A sinusoidal wave with validated amplitude and frequency.
final class Wave {
  /// Creates a wave with an optional phase offset in radians.
  Wave(this.amplitude, this.frequency, [this.phase = 0.0]) {
    if (!amplitude.isFinite || amplitude < 0.0) {
      throw ArgumentError.value(
        amplitude,
        'amplitude',
        'must be finite and non-negative',
      );
    }
    if (!frequency.isFinite || frequency <= 0.0) {
      throw ArgumentError.value(
        frequency,
        'frequency',
        'must be finite and positive',
      );
    }
    if (frequency > double.maxFinite / trig.twoPi) {
      throw ArgumentError.value(
        frequency,
        'frequency',
        'must keep angular frequency finite',
      );
    }
    if (!phase.isFinite) {
      throw ArgumentError.value(phase, 'phase', 'must be finite');
    }
  }

  /// Peak displacement from equilibrium.
  final double amplitude;

  /// Complete cycles per second.
  final double frequency;

  /// Initial phase offset in radians.
  final double phase;

  /// Duration of one complete cycle in seconds.
  double period() => 1.0 / frequency;

  /// Phase-change rate in radians per second.
  double angularFrequency() => 2.0 * trig.pi * frequency;

  /// Evaluates `amplitude * sin(2 * pi * frequency * time + phase)`.
  double evaluate(double time) {
    if (!time.isFinite) {
      throw ArgumentError.value(time, 'time', 'must be finite');
    }
    if (amplitude == 0.0) {
      return 0.0;
    }

    // Reduce time to one period before multiplying. This keeps the phase
    // finite even when two individually finite inputs would overflow if they
    // were multiplied directly.
    final representedPeriod = period();
    final reducedTime =
        representedPeriod.isInfinite ? time : time % representedPeriod;
    final reducedPhase = phase % trig.twoPi;
    final angle = trig.twoPi * (frequency * reducedTime) + reducedPhase;
    final unitValue = trig.sin(angle);
    if (unitValue >= 1.0) {
      return amplitude;
    }
    if (unitValue <= -1.0) {
      return -amplitude;
    }
    return amplitude * unitValue;
  }

  @override
  String toString() =>
      'Wave(amplitude: $amplitude, frequency: $frequency, phase: $phase)';
}
